docs/design/values.md

```diff
@@ -0,0 +1,917 @@
+_storage_ where values can be read and written. Storage also allows taking the
+address of an object in memory in Carbon.
+
+Both objects and values can be nested within each other. For example
+`(true, true)` is both a value and also contains two sub-values. When a
+two-tuple is stored somewhere, it is both a tuple-typed object and contains two
+subobjects.
+
+These terms are important components in the describing the semantics of Carbon
+code, but they aren't sufficient. We also need to explicitly and precisely talk
+about the Carbon _expressions_ that produce or reference values and objects.
+Categorizing the expressions themselves allows us to be more precise and
+articulate important differences not captured without looking at the expression
+itself.
+
+### Expression categories
+
+There are three expression categories in Carbon:
+
+-   [_Value expressions_](#value-expressions) produce abstract, read-only
+    _values_ that cannot be modified or have its address taken.
+-   [_Reference expressions_](#reference-expressions) refer to _objects_ with
+    _storage_ where a value may be read or written and the object's address can
+    be taken.
+    -   [_Durable reference expressions_](#durable-reference-expressions) are
+        reference expressions which cannot refer to _temporary_ storage, but
+        must refer to some storage that outlives the full expression.
+    -   [_Ephemeral reference expressions_](#ephemeral-reference-expressions)
+        are reference expressions which _can_ refer to temporary storage.
+-   [_Initializing expressions_](#initializing-expressions) which require
+    storage to be provided implicitly when evaluating the expression. The
+    expression then initializes an object in that storage. These are used to
+    model function returns, which can construct the returned value directly in
+    the caller's storage.
+
+The general conversions and the semantics implied between these categories are
+below:
+
+-   An _initializing expression_ can be formed from:
+    -   A _value expression_ by using the value to initialize an object in the
+        provided storage, analogous to
+        [direct initialization](https://en.cppreference.com/w/cpp/language/direct_initialization)
+        in C++.
+    -   A _durable reference expression_ by copying the referenced object into
+        the new storage, analogous to
+        [copy initialization](https://en.cppreference.com/w/cpp/language/copy_initialization)
+        in C++.
+-   An _ephemeral reference expression_ can be formed from:
+    -   A _durable reference expression_ trivially.
+    -   An _initializing expression_ by materializing temporary storage for an
+        object that is initialized and then referencing that object.
+-   A _durable reference expression_ cannot be produced by converting from
+    another expression category.
+-   A _value expression_ can be formed from a _reference expression_ by reading
+    the value from its storage.
```

Comment:
I agree this text is misleading and it would be more accurate to link to the bit where the configuration of value representation for a type is described, possibly with a short summary here -- or even "this may involve reading the value from its storage"

Reply:
We now have updated text and an updated name!

I'm trying out "value binding" here and saying that the value expression *binds* an object in storage.

The goal is to get at the idea that mutation breaks that binding and that the two are linked until broken.

Mini brain-storming didn't come up with any better terminology than this. While it's a bit unfortunate to overlap a bit with *name* binding, I actually don't mind it, and in fact binding a name to a value also often binds that value to an object.

(Or should it bind the object to the value? Interested in what the best directionality is here.)